### PR TITLE
Add REX learning rate scheduler

### PR DESCRIPTION
**Pull request by xzuyn**
This adds the REX learning rate scheduler from [**REX: Revisiting Budgeted Training with an Improved Schedule**](https://arxiv.org/abs/2107.04197). The implementation was modified from [IvanVassi/REX_LR](https://github.com/IvanVassi/REX_LR).

Use `--lr_scheduler rex` to choose it. Supports `--lr_scheduler_min_lr_ratio` and `--lr_warmup_steps`.

`--lr_scheduler_min_lr_ratio 0.01` is used by default, as without it the first step (if using warmup) and last step will use 0 learning rate.

Example with `--lr_scheduler rex --learning_rate 1e-4 --lr_scheduler_min_lr_ratio 0.01 --lr_warmup_steps 10 --max_train_steps 100`, comparing the paper's values (`0.5+0.5`) and the values IvanVassi's implementation used (`0.1+0.9`). This PR is using `0.1+0.9`.
rex-comparison

### CODE REVIEW

**Comment by kohya-ss** [comment]
Thank you for this PR! Could you please tell me the difference when using the polynomial scheduler with power<1.0?

It seems to draw a curve very similar to REX.

Figure_1

**Comment by xzuyn** [comment]
They are similar, but not the same. REX lets you set the `min_lr` which is respected for warm-up first step and also the last training step. Polynomial starts and ends on two different values with no option for a `min_lr`, and does a sharp drop on the last step.

REX args: `--lr_scheduler rex --learning_rate 1e-4 --lr_scheduler_min_lr_ratio 0 --lr_warmup_steps 10 --max_train_steps 100`
Polynomial args: `--lr_scheduler polynomial --learning_rate 1e-4 --lr_scheduler_power [0.25, 0.5, 0.75] --lr_warmup_steps 10 --max_train_steps 100`
Screenshot 2025-09-03 at 02-53-36 Qwen-Image-musubi Workspace – Weights   Biases

**Comment by kohya-ss** [comment]
Thanks for the clarification. Honestly, I don't think there's much difference...

I wonder if it would be possible to make the two values ​​related to rex_factor into parameters, so that users could choose between the paper implementation, IvanVassi's, or something else. That might be a useful alternative to Polynomial.

**Comment by xzuyn** [comment]
You can now use `--rex_alpha` and `--rex_beta` to adjust the values. It uses IvanVassi's values by default. Let me know if you would rather it use the paper's values, or use different variable names.

**Comment by kohya-ss** [comment]
Thank you! I will review and merge this sooner!

**Comment by kohya-ss** [comment]
Thank you again for this.

It is not desirable to add parameters that depend on a specific scheduler to the training script, so after merging, I would like to change `rex_alpha` and `rex_beta` to be obtained from `--lr_scheduler_args`. I will also move the class to a separate module. Thank you for your understanding.

**Comment by xzuyn** [comment]
Oh I didn't even see that arg. That would have been better to use.
